sdk/dotnet/Pulumi.Automation/PublicAPI.Shipped.txt

```diff
@@ -170,6 +170,8 @@ Pulumi.Automation.PreviewOptions.Program.get -> Pulumi.Automation.PulumiFn
 Pulumi.Automation.PreviewOptions.Program.set -> void
 Pulumi.Automation.PreviewOptions.Replace.get -> System.Collections.Generic.List<string>
 Pulumi.Automation.PreviewOptions.Replace.set -> void
+Pulumi.Automation.PreviewOptions.SerilogLoggerFactory.get -> System.Func<Serilog.ILogger>
```

Comment:
Hmm :) @mikhailshilkov do we have a policy?

I've worked on projects before that were very conservative about this, but I get a feeling we are a bit more liberal here. Maybe we can accept this?

Are there hacks that would prevent Serilog surfacing in the interface? Such as taking opaque string/object settings map and recovering via reflection?

Reply:
Note that taking a dependency on the [Pulumi](https://www.nuget.org/packages/Pulumi) NuGet package subsequently takes a dependency on the [Serilog](https://www.nuget.org/packages/Serilog/) NuGet package and its siblings. It's not used as developer-time only dependency, so surfacing a type from this dependency seems reasonable and in this case is also needed to empower the hosting application to configure it accordingly.

Also, it may not be clear since I sorted the [package references](https://github.com/pulumi/pulumi/blob/cb539c58d2fad759d1f20f8923fce877ae3b79a8/sdk/dotnet/Pulumi/Pulumi.csproj#L33-L48) for readability, but I also removed references to [Microsoft.Extensions.Configuration.EnvironmentVariables](https://www.nuget.org/packages/Microsoft.Extensions.Configuration.EnvironmentVariables), [Serilog.Settings.Configuration](https://www.nuget.org/packages/Serilog.Settings.Configuration) and [Serilog.Sinks.File](https://www.nuget.org/packages/Serilog.Sinks.File) since that would depend on and be fully configured by the host, which can even be leveraged by test hosts as shown [here](https://github.com/pulumi/pulumi/blob/cb539c58d2fad759d1f20f8923fce877ae3b79a8/sdk/dotnet/Pulumi.Automation.Tests/LocalWorkspaceTests.cs#L58-L61).

Reply:
@gitfool I don't think the fact that `Pulumi` has a reference to `Serilog` alone is argument enough to expose a type on `Pulumi`'s API surface tying it to `Serilog`. Pulumi can change what they use for logging at any point without it being a breaking change if they are not exposing it on their public API, but by tying Serilog to their public API changing what package they use for logging becomes a breaking change.

But like I said, I defer to the pulumi team here.

Reply:
@orionstudt points taken and agreed. I was just thinking it's reasonable given the credentials of Serilog and that an abstraction here is not worth the effort.

Reply:
Chatting with @mikhailshilkov we've reached a conclusion we do care about this and would like to consider some alternatives here if possible. Such as wrapping the config in our own type (if it's possible to isolate reasonable properties that do not seem fully committing this type to Serilog specifics), or even having an "LogConfiguration System.Object" passed in and detecting Serilog by reflection?

Reply:
I propose we bring this back high-level and do something a lot lighter weight. I propose that we expose some variation of the `Pulumi.ILogger` interface, and we modify the appropriate `XOptions` classes to take in an optional `Pulumi.ILogger?` instance.

If it is not provided, we default it to our internal Serilog implementation that reads just the environment config that @gitfool implemented in #7095. If it is provided by the consumer, than we use their implementation. This way if the consumer wants complex Serilog config they can just write their own implementation of `Pulumi.ILogger` that does whatever we want, and we also support all different logging libraries into the future.

Reply:
After thinking about it some more we'd probably want the API surface of an exposed `ILogger` interface to just take in a `string message` and a `LogLevel level` so that the consumer isn't receiving a `Pulumi.Resource` instance that could be used to potentially affect the ongoing update.

Reply:
Okay, in that case I think a much better abstraction for "external loggers" would be to use [Microsoft.Extensions.Logging.Abstractions](https://www.nuget.org/packages/Microsoft.Extensions.Logging.Abstractions), specifically [ILogger\<TCategoryName\>](https://github.com/dotnet/runtime/blob/bdc5d79e9976aeeeabe2a6e38b1e80e33bd01506/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/ILoggerT.cs#L12), given Serilog maps `TCategoryName` to its source context; see also [ILogger](https://github.com/dotnet/runtime/blob/bdc5d79e9976aeeeabe2a6e38b1e80e33bd01506/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/ILogger.cs#L12) and [LoggerExtensions](https://github.com/dotnet/runtime/blob/bdc5d79e9976aeeeabe2a6e38b1e80e33bd01506/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerExtensions.cs#L11). That way support for all logging libraries is already available.

This could be a drop in replacement for the Serilog logger and kept separate from the Pulumi logger, as it is now, so that the Pulumi logger controls what flows through the "external logger".

Reply:
Ok #7246 is my proposal. @gitfool your suggestion to allow consumer to provide a `Microsoft.Extensions.Logging.ILogger` is a good one though so I will modify my PR to do that.

I also left out the log enrichment changes you did so that you could do them in a separate PR

Reply:
@orionstudt I was in the middle of adapting my PR. I've just pushed another commit. Take a look.

Reply:
Ok, whatever works. I have some thoughts after writing it myself:
- I think it should just be an `ILogger` instance instead of an `ILoggerProvider`. We're only going to invoke the provider once and even if we weren't we're giving it the same category every time so it's moot I think. Also by making it `ILogger` the consumer can control the category
- I think the property should just be `Logger` - if the only logger a consumer sees is `ExternalLogger` than I don't think the `External` distinction is necessary
- I think the property should also be on `LocalWorkspaceOptions` and `Workspace` the same way that each of those have `Program`. And allow propagation down from there the same way `Workspace.Program` propagates that way consumer can just provide the logger on the workspace instead of needing to provide it on every update action.
- Everywhere the consumer can provide an `ILogger` we should have a remark that it will only be used if `Program` is also provided, i.e. for inline-programs
- I don't think we need to pollute the other unit tests to test it, we can just have a single independent test verifying that the provided `ILogger` is invoked for each of the 3 ways it can be provided, `UpOptions`, `PreviewOptions`, & `LocalWorkspaceOptions`

Reply:
Done

Reply:
Really sorry swamped with a few P1s currently, there will be a day delay here.